x/ecocredit/server/basket/msg_put_test.go

```diff
@@ -0,0 +1,478 @@
+		Owner:       s.alice.String(),
+		BasketDenom: s.basketDenom,
+		Credits: []*basket.BasketCredit{
+			{
+				BatchDenom: a,
+				Amount:     s.tradableCredits,
+			},
+		},
+	})
+}
+
+func (s *putSuite) BobAttemptsToPutCreditsFromCreditBatchIntoTheBasket(a string) {
+	amount, _ := sdk.NewIntFromString(s.tradableCredits)
+	coins := sdk.NewCoins(sdk.NewCoin(s.basketDenom, amount))
+
+	s.bankKeeper.EXPECT().
+		MintCoins(s.sdkCtx, basket.BasketSubModuleName, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	s.bankKeeper.EXPECT().
+		SendCoinsFromModuleToAccount(s.sdkCtx, basket.BasketSubModuleName, s.bob, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	_, s.err = s.k.Put(s.ctx, &basket.MsgPut{
+		Owner:       s.bob.String(),
+		BasketDenom: s.basketDenom,
+		Credits: []*basket.BasketCredit{
+			{
+				BatchDenom: a,
+				Amount:     s.tradableCredits,
+			},
+		},
+	})
+}
+
+func (s *putSuite) AliceAttemptsToPutCreditsIntoTheBasket() {
+	amount, _ := sdk.NewIntFromString(s.tradableCredits)
+	coins := sdk.NewCoins(sdk.NewCoin(s.basketDenom, amount))
+
+	s.bankKeeper.EXPECT().
+		MintCoins(s.sdkCtx, basket.BasketSubModuleName, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	s.bankKeeper.EXPECT().
+		SendCoinsFromModuleToAccount(s.sdkCtx, basket.BasketSubModuleName, s.alice, coins).
+		Return(nil).
+		AnyTimes() // not expected on failed attempt
+
+	_, s.err = s.k.Put(s.ctx, &basket.MsgPut{
+		Owner:       s.alice.String(),
+		BasketDenom: s.basketDenom,
+		Credits: []*basket.BasketCredit{
+			{
+				BatchDenom: s.batchDenom,
+				Amount:     s.tradableCredits,
+			},
+		},
+	})
+}
+
+func (s *putSuite) TheBasketHasACreditBalanceWithAmount(a string) {
+	basket, err := s.stateStore.BasketTable().GetByBasketDenom(s.ctx, s.basketDenom)
+	require.NoError(s.t, err)
+
+	balance, err := s.stateStore.BasketBalanceTable().Get(s.ctx, basket.Id, s.batchDenom)
+	require.NoError(s.t, err)
+
+	require.Equal(s.t, a, balance.Balance)
+}
+
+func (s *putSuite) TheBasketTokenHasATotalSupplyWithAmount(a string) {
+	basket, err := s.stateStore.BasketTable().GetByBasketDenom(s.ctx, s.basketDenom)
+	require.NoError(s.t, err)
+
+	amount, err := strconv.ParseInt(a, 10, 32)
+	require.NoError(s.t, err)
+
+	coin := sdk.NewInt64Coin(s.basketDenom, amount)
+
+	s.bankKeeper.EXPECT().
+		GetSupply(s.sdkCtx, basket.BasketDenom).
+		Return(coin).
+		Times(1)
+
+	supply := s.bankKeeper.GetSupply(s.sdkCtx, s.basketDenom)
+	require.NotNil(s.t, supply)
+	require.Equal(s.t, coin, supply)
+}
+
+func (s *putSuite) AliceHasACreditBalanceWithAmount(a string) {
+	batch, err := s.coreStore.BatchTable().GetByDenom(s.ctx, s.batchDenom)
+	require.NoError(s.t, err)
+
+	balance, err := s.coreStore.BatchBalanceTable().Get(s.ctx, s.alice, batch.Key)
+	require.NoError(s.t, err)
+
+	require.Equal(s.t, a, balance.Tradable)
+}
+
+func (s *putSuite) AliceHasABasketTokenBalanceWithAmount(a string) {
```

Comment:
Same as above. Each of these functions are not tests but steps in a test scenario. I can add another step to the scenario that this step is being used within and the new step would check the response to make sure the response was set. When alice attempts to put credits into the basket, we can set both the response and the error and then check both, but this would be a new step (a new function) and we wouldn't update this step function.

Reply:
I could also add a new scenario that would only check the response.

Reply:
I think the steps are fine, I would just rather this specific part test the amount of coins in the response. That way we are actually capturing more surface area for `Put`'s behavior to catch potential errors in its implementation. the way this specific step is implemented does not really capture any surface area for `Put` since the bank keeper is mocked.

Reply:
Already added a step for checking the response. https://github.com/regen-network/regen-ledger/pull/1065/commits/7226d60290d132906e00fcd3fb85ff7d16ad08d0

Reply:
[resolving](https://github.com/regen-network/regen-ledger/pull/1065#discussion_r867009366)